[18.0][MIG] stock_picking_group_by_partner_by_carrier

theangryangel opened the pull request:
I needed to make some assumptions and adjustments to this module based on how 18.0 changes the standard / out-of-the-box multi-step routes from pull to push routes.

Notable changes:
* The module now supports both push and pull rule configuration. When in the new-style push rules, the behaviour is slightly different.
* Now depends on `stock_warehouse_out_pull` to easily/quickly accommodate the fact that historical database may be configured with pull rules and new database will have push rules.
* `group_pickings` is now visible based on the configuration from stock_warehouse_out_pull - if push rules then it's shown as per previous behaviour. If it's new pull rules then it's shown at all times.
* `StockMove._domain_search_picking_handle_move_type` required an additional if statement added to ensure that `TestMergeWizard` passed all tests
* `StockPicking._merge_procurement_groups` required a change in the from `picking_type_id.code != "outgoing"` to checking the `picking_type_id.group_pickings`
* Switched from itertools.groupby to odoo.tools.groupby
* New tests were added to handle the push rules
* Old test file handling pull rules was renamed for clarity
* `sale_stock` was added as a dependency and the delivery slip report inherit_id changed as it was depending on a field added through `sale_stock`

gurneyalex committed:
[ADD] stock_picking_group_by_partner_by_carrier

This module can be used if your customers expect that several different orders
they passed will be shipped in a single delivery order.

With this module installed, when a sale order is confirmed, the stock moves for
the lines of the sale order can be placed in an existing delivery order that
shares the same delivery address and carrier (or lack thereof).

Sale orders with a Shipping Policy set to 'When all products are ready' always
get their own shipping.

gurneyalex committed:
[FIX] cancellation of sale order and WH multi step

* the grouped picking must not be cancelled

* in WH multi step mode:
** ensure the picking ids are computed correctly
** the sale order cancelation does what is expected

sebalix committed:
[FIX] group pick operations as well with pick-ship delivery steps

sebalix committed:
[FIX] stock_picking_group_by_partner_by_carrier: add a hook to handle domain on move_type

And add a test with the 'pick_pack_ship' configuration with grouping
enabled on pack+ship (but not pick).

TDu committed:
[13.0][IMP] stock_picking_group_by_partner_by_carrier

Improve the generated delivery note report. By adding the customer reference
of the sale order on each section.

grindtildeath committed:
stock_picking_group_by_partner_by_carrier: Display carrier on PG

Commit:
Fix groups of transfers on sales

When grouping transfers of different sales orders in the same
procurement.group, we have an issue with only the first sales order
showing the "Transfers" button on the SO view, because
"sale.picking_ids" returns an empty recordset on the other sales orders.

This is because the link between sales and pickings is done through the
"group_id.sale_id" field on the stock moves. When we group transfers of
different sales, we end up with several group_id in the moves of the
same picking, which is a bit unexpected, and when new transfers are
created (dynamic routing, backorders, ...), they'll only get the first
group_id and be attached only to the first sale order.

The solution implemented here is the following:

* Add a M2m "sale_ids" field on the procurement group
* When several moves with different groups are merged into the same
  group, clone the group, update its "sale_ids" field to contain all the
  sales (sale_id will keep the first one though) and set in on the moves
* Update the computed fields for the relation between the sales and
  pickings to use this M2M field
* Keep the original group in the stock move as "original_group_id": used
  when a sales order is cancelled

Commit:
Simplify line group comparison

Commit:
Prevent moves being always merged in the same groups

If we have a unique transfer for 2 sales orders, we have one joint
procurement group for them. If we delivery the first sale order, we
expect to no longer have the first SO in the group of the backorder.

This will prevent to have a never-ending group with all the orders from
the beginning if we have many backorders.

To so do, we consider that a "printed" or done transfer will no longer
change and therefore, we consider its group frozen.

Commit:
Rename/move around methods for clarity

TDu committed:
[13.0][IMP] stock_picking_group_by_partner_by_carrier

The delivery slip can reference more than one sale order with different
customer references. If there is more than one of those references they
will be displayed in the lines.
But only one in the header, better not to display the customer reference
in the header with multiple sale order.
Also fix the customer reference to be specific to the delivery lines.

jbaudoux committed:
Fix cancelation of SO when multiple SO are grouped in a same delivery.

Propagate the original_group_id on pull moves to allow to prevent
merging pulled moves. This allows to cancel a SO without canceling
pulled moves from other SO, as we ensure they are not merged.

squash! Fix cancelation of SO when multiple SO are grouped in a same delivery.

Add tests

AaronHForgeFlow committed:
[FIX]stock_picking_group_by_partner_by_carrier
* add valid development status

TDu committed:
[13.0][IMP] stock_picking_group_by_partner_by_carrier

Remove the sentence about remaining items to be shipped.
As it is confusing with multiple sale order in a delivery

simahawk committed:
stock_picking_group_by_* rebase on stock_move_assign_picking_hook

simahawk committed:
s_picking_group_by_partner_by_carrier: add merge wizard

sebalix committed:
s_picking_group_by_partner_by_carrier: add a regression test when creating backorder

If the group pickings option is disabled, the creation of a backorder
when validating a partial qty of a picking should work as usual.

sebalix committed:
s_picking_group_by_partner_by_carrier: fix regression when creating backorder

TDu committed:
[13.0][FIX] stock_picking_group_by_partner_by_carrier

When a sale order is cancelled then set to quotation again, all
related picking are cancelled also.

If before reconfirming the sale order for a 2nd time the delivery
carrier is changed, the procurement group linked on the sale order
will be used and related to the wrong carrier.

Removing the assigned procurement group on the reactivation
of the sale order fixes the issue.

TDu committed:
s_picking_group_by_partner_by_carrier: Fix MockedMove

Using an alternative data structure to store the separation of the
sale order on the delivery note as been sensible to breakage from
changes upstream.

Using the same type of records than the rest of the information
will hopefully help.

phuctranfxvn committed:
[IMP] stock_picking_group_by_partner_by_carrier: black, isort, prettier

Commit:
[IMP] stock_picking_group_by_partner_by_carrier

When the delivery slip is printed, the list of pending quantities
to deliver is shown at the end, grouped by order.

phuctranfxvn committed:
[MIG] stock_picking_group_by_partner_by_carrier: Migration to 14.0

sebalix committed:
s_picking_group_by_*: disable group on partner option

Allow to disable the grouping feature on partner and fallback on the
standard Odoo behavior.

sebalix committed:
[FIX] s_p_g_by_partner_by_carrier: remaining to deliver

Several moves could be related to a sale order line, qty should be
computed from all of them.

+ regarding the delivered qties, print only the moves having some
qty reserved if the transfer is not yet done.

mmequignon committed:
[13.0][IMP] s_p_g_by_partner_by_carrier: Add hook on delivery report

mmequignon committed:
s_p_group_by_partner_by_carrier: Fix translations in report

sebalix committed:
[IMP] s_p_g_by_p_by_carrier: hook to check state

sebalix committed:
[IMP] s_p_group_by_partner_by_carrier: black, isort, prettier

sebalix committed:
[FIX] s_p_group_by_partner_by_carrier: fix merge wizard

jbaudoux committed:
[FIX] Properly merge the procurement groups

In case of a inter-WH resupply delivery order (i.e. not linked to any
SO), the grouping was effective but the procurement group was not merged
properly.

In case of SO, when adding new lines, the procurement group on the new
line was not correct.

In multi-step delivery, the new procurement group was not propagated all
to the new internal moves.

Do not create a new procurement group when not necessary and a rename
can be applied.

Sort procurement group name and origin by name.

Post message on picking that it has been modified when new moves from SO
are added (instead of re-posting that it has been created).

Also update origin in case of inter-WH resupply.

sebalix committed:
[FIX] s_p_group_by_: fix printing of delivery slip

Two bugfixes:
- filter out canceled moves
- in case there is no sale order related to some moves (even if this should
  not happen), do not print an empty section line

TDu committed:
Add hooks on stock_picking_group_by_partner_by_carrier

This adds two hooks to enable on the report:

* An alternative qantity calculation
* Easy customization of sales data

jbaudoux committed:
Add procurement_group_carrier. Extract feature from stock_picking_group_by_partner_by_carrier

simahawk committed:
s_picking_group_by_partner_by_carrier: dev status -> Beta

mmequignon committed:
Added translation using Weblate (German)

mmequignon committed:
Added translation using Weblate (French)

mmequignon committed:
Added translation using Weblate (Italian)

mmequignon committed:
s_p_group_by_partner_by_carrier: Fix deliveryslip report

https://github.com/odoo/odoo/blob/ef0320b82e679f2f6eb4b2ba8aa4530cda9a98a9/addons/stock/report/report_deliveryslip.xml#L77-L124

There's 4 different cases for the deliveryslip report:
 1) package and serial number
 2) package and no serial number
 3) no package but serial number
 4) no package and no serial number

For the 2nd an 4th case, odoo tries to aggregate the lines and therefore
shouldn't be groupped by SO, no fix needed here.

However, when there's serial numbers, we can group lines by SO, and we
should use `picking.get_delivery_report_lines()` to get the lines to
display. The 3rd case wasn't handled yet.

Also, when groupping by SO, the `get_delivery_report_lines` creates fake
`stock.move.line` for each SO section, with product_id = 1.
However, odoo tries to display the product, which is wrong for the
SO sections.

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 5.7% (3 of 52 strings)

Translation: stock-logistics-workflow-14.0/stock-logistics-workflow-14.0-stock_picking_group_by_partner_by_carrier
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-14-0/stock-logistics-workflow-14-0-stock_picking_group_by_partner_by_carrier/it/

rousseldenis committed:
[14.0][FIX] Align development statuses

rousseldenis committed:
[MIG][16.0] stock_picking_group_by_partner_by_carrier

rousseldenis committed:
[IMP] stock_picking_group_by_partner_by_carrier: printed field is copy=False in core

rousseldenis committed:
[IMP] stock_picking_group_by_partner_by_carrier: Use attirbutes instead of replace

rousseldenis committed:
[IMP] stock_picking_group_by_partner_by_carrier: Add index on grouping fields

glitchov committed:
[IMP] stock_picking_group_by_partner_by_carrier : fr translations

lmignon committed:
[IMP] stock_picking_group_by_partner_by_carrier: Add index on picking.group_id

In the '_merge_procurement_groups' from the 'stock.picking' model, pickings are browsed from the group_id field of a stock_move ('group_pickings = self.move_ids.group_id.picking_ids.filtered'. We add an index on the 'group_id' field on the 'stock_picking' model to make this process efficient on large database

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (51 of 51 strings)

Translation: stock-logistics-workflow-16.0/stock-logistics-workflow-16.0-stock_picking_group_by_partner_by_carrier
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-16-0/stock-logistics-workflow-16-0-stock_picking_group_by_partner_by_carrier/it/

sbejaoui committed:
[FIX] stock_picking_group_by_partner_by_carrier: simplify the name of merged procurment groups

Prior to this change, we observed records in our database with names exceeding 1GB...

EmilioPascual committed:
[FIX] stock_picking_group_by_partner_by_carrier: Added condition to check get_remaining_to_deliver

In some times, method get_remaining_to_deliver is not defined in template.

@moduon MT-4407

Ivorra78 committed:
Added translation using Weblate (Spanish)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (53 of 53 strings)

Translation: stock-logistics-workflow-16.0/stock-logistics-workflow-16.0-stock_picking_group_by_partner_by_carrier
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-16-0/stock-logistics-workflow-16-0-stock_picking_group_by_partner_by_carrier/it/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (53 of 53 strings)

Translation: stock-logistics-workflow-16.0/stock-logistics-workflow-16.0-stock_picking_group_by_partner_by_carrier
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-16-0/stock-logistics-workflow-16-0-stock_picking_group_by_partner_by_carrier/es/

rousseldenis committed:
[FIX] stock_picking_group_by_partner_by_carrier: Don't get the assigned picking group

As Odoo changed its behavior to auto assign picking group to moves when picking_id field
is filled in (backorder creation, assignation), we let the original group on move level.

rousseldenis committed:
[IMP] stock_picking_group_by_partner_by_carrier: Restore ignored test

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (53 of 53 strings)

Translation: stock-logistics-workflow-16.0/stock-logistics-workflow-16.0-stock_picking_group_by_partner_by_carrier
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-16-0/stock-logistics-workflow-16-0-stock_picking_group_by_partner_by_carrier/it/

RodrigoBM committed:
[FIX] stock_picking_group_by_partner_by_carrier: add the same class style as in the base report

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (53 of 53 strings)

Translation: stock-logistics-workflow-16.0/stock-logistics-workflow-16.0-stock_picking_group_by_partner_by_carrier
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-16-0/stock-logistics-workflow-16-0-stock_picking_group_by_partner_by_carrier/it/

samibc2c committed:
Translated using Weblate (French)

Currently translated at 96.2% (51 of 53 strings)

Translation: stock-logistics-workflow-16.0/stock-logistics-workflow-16.0-stock_picking_group_by_partner_by_carrier
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-16-0/stock-logistics-workflow-16-0-stock_picking_group_by_partner_by_carrier/fr/

theangryangel committed:
[IMP] stock_picking_group_by_partner_by_carrier: pre-commit auto fixes

theangryangel committed:
[MIG] stock_picking_group_by_partner_by_carrier: Migration to 18.0

jbaudoux commented on stock_picking_group_by_partner_by_carrier/models/stock_move.py:
```diff
@@ -114,8 +111,11 @@ def _domain_search_picking_handle_move_type(self):
-        return [("move_type", "=", self.group_id.move_type)]
+        if self.group_id.move_type:
+            # avoid mixing picking policies
+            return [("move_type", "=", self.group_id.move_type)]
```
if the move type is missing on the procurement group, it now defaults to the one configured on the picking type. Otherwise, we can assume direct as default.
```suggestion
        group = self.group_id.move_type or self.picking_type_id.move_type or "direct"
        return [("move_type", "=", self.group_id.move_type)]
```

jbaudoux commented on stock_picking_group_by_partner_by_carrier/models/stock_picking.py:
```diff
@@ -111,21 +111,21 @@ def _update_merged_origin(self):
                 partners_name=", ".join(partners.mapped("display_name")),
             )
-        return {"sale_ids": [(6, 0, sales.ids)], "name": name}
+        return {"sale_ids": [(6, 0, sales.ids)], "name": name, "sale_id": False}
```
Do not include https://github.com/OCA/stock-logistics-workflow/pull/1789 in your migration commit.
And as I explained, this is a wrong approach as it creates incompatibility with all modules relying on this standard `sale_id` value.

theangryangel replied:
Reverted.

jbaudoux commented on stock_picking_group_by_partner_by_carrier/report/report_delivery_slip.xml:
```diff
@@ -167,13 +168,13 @@
         <!-- Replace the customer refs with the one based on the move line displayed -->
         <xpath
-            expr='//p[@t-field="o.sudo().sale_id.client_order_ref"]'
+            expr='//div[@t-field="o.sudo().sale_id.client_order_ref"]'
```
Not linked to the migration but shouldn't the template inherit `sale_stock.report_delivery_document_inherit_sale_stock` instead ?
If you do this change, do it in a separate fix commit

jbaudoux commented on stock_picking_group_by_partner_by_carrier/views/stock_picking_type.xml:
```diff
@@ -5,13 +5,9 @@
-                />
-            </field>
+            <xpath expr="//group[@name='first']/group[2]" position="inside">
+                <field name="group_pickings" />
```
This change needs a big warning in the module readme.
You can configure grouping of picking only in case of push flow. Otherwise, only the outgoing can be enabled

theangryangel replied:
I've added some additional logic to show/hide based on stock_warehouse_out_pull and I've added some extra information to the documentation describing the difference in configuration. Not sure if it's a strong enough message though.

jbaudoux reviewed (changes requested):
Thanks for the work.

It's nice you added tests for the new push flow of odoo but we need to keep the tests for the case you continue using pull flow to ensure non regression.
You can duplicate the impacted tests. And you can rely on https://github.com/OCA/stock-logistics-warehouse/pull/2353 for applying pull flow setup.

jbaudoux commented:
@theangryangel your changes to support the new push flow are legit and I don't see any change breaking the previous behavior. Sounds good to me. Keeping also the previous tests as they were should show there are no regressions.
If you want to use https://github.com/OCA/stock-logistics-workflow/pull/1789, put it in a separate PR as it creates incompatibilities and won't be merged as is.

theangryangel commented:
>  I don't see any change breaking the previous behavior. Sounds good to me. Keeping also the previous tests as they were should show there are no regressions.

I had significant issues trying to support both workflows. It might've been that I was having a blurry-eyed halucination from a 2am wake up from a toddler 😅 

I'll take a look at some point today or tomorrow and re-add those tests :)

Thank you for the feedback <3

theangryangel committed:
[FIX] stock_picking_group_by_partner_by_carrier: improper template inherit_id

theangryangel commented:
I believe that I have made the relevant changes. I've incorporated a fix from bizappdev and the suggestions from jbaudoux. I've updated the PR description with the current state of things.

I'm not sure if the USAGE and CONFIGURATION changes are clear enough. Input would be appreciated.

jbaudoux commented on stock_picking_group_by_partner_by_carrier/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,21 @@
+
+Note: The grouping is currently not propagated to the pulled internal
+moves if you don't use the module stock_available_to_promise_release in
+the wms repository.
```
it's being moved
```suggestion
the stock-logistics-workflow repository.
```

jbaudoux reviewed (approved):
Thanks for the updates

can you rewrite your fix commit to add the module name?
`[FIX] stock_available_to_promise_release: report inherit_id`

bizzappdev commented on stock_picking_group_by_partner_by_carrier/__manifest__.py:
```diff
@@ -0,0 +1,35 @@
+    "category": "Warehouse",
+    "depends": [
+        "delivery_procurement_group_carrier",
+        "sale_stock",
```
IMO, the stock_delivery module https://github.com/odoo/odoo/blob/18.0/addons/stock_delivery/__manifest__.py#L14 is already adding the dependency, explicit dependency is unnecessary.

simahawk replied:
It depends. If we have a direct dependency on stock_delivery (a field, an override, a view, whatever), it's better to keep it explicit.

theangryangel replied:
I believe that I have followed the rules in terms of squashing and reordering. Thank you for pointing it out bizzappdev. I'd completely forgotten that step. It's been a moment since I last contributed :/

bizzappdev reviewed (changes requested):
@theangryangel Thank you for the contribution,
Please squash administrative commits (if any) with the previous commit for reducing commit noise as per the [migration checklist](https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-18.0) requirements. Ref: [Merge/squash the commits generated by bots or Weblate](https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests#mergesquash-the-commits-generated-by-bots-or-weblate)

theangryangel commented:
I believe I've made all the requested changes. Anything else, please do let me know :)

jbaudoux commented on stock_picking_group_by_partner_by_carrier/models/stock_move.py:
```diff
@@ -0,0 +1,128 @@
+        if (
+            not self.picking_type_id.group_pickings
+            or self.partner_id.disable_picking_grouping
+            or self.group_id.sale_id.picking_policy == "one"
```
Not grouping when `one` should be configurable

jbaudoux replied:
improved in https://github.com/OCA/stock-logistics-workflow/pull/1987

jbaudoux commented on stock_picking_group_by_partner_by_carrier/models/stock_move.py:
```diff
@@ -0,0 +1,128 @@
+            ("partner_id", "=", self.group_id.partner_id.id),
+            # don't search on the procurement.group
+        ]
+        domain += self._domain_search_picking_handle_move_type()
```
When the move type is `one`, the deadline should be respected when there is a `sale_line_id.order_id.commitment_date`

theangryangel replied:
So I'll counter that one - The customer we're porting this for actually specifically ignores the commitment date if they can consolidate, risking upsetting the customer by delaying one order by a day or two, but proactively inform the customer that they're gonna get order A, B and at once, for example.

The customer doesn't really get a choice, the margins of the business are "low".

I'd look to make this configurable, but I'm hard pressed to devote time to add additional features on this PR right now

jbaudoux replied:
@theangryangel See https://github.com/OCA/stock-logistics-workflow/pull/1987

jbaudoux commented on stock_picking_group_by_partner_by_carrier/models/stock_move.py:
```diff
@@ -0,0 +1,128 @@
+            domain.append(("id", "!=", self.picking_id.id))
+        return domain
+
+    def _domain_search_picking_handle_move_type(self):
```
What we want to hook is the retrieval of the `move_type`, not the domain

jbaudoux reviewed (commented):
Not linked to the migration, but would be nice to allow merging also when the policy is `one` unless there is a commitment date

bizzappdev reviewed (approved):
Code and Funtional test LGTM

simahawk commented:
/ocabot migration stock_picking_group_by_partner_by_carrier

simahawk commented:
/ocabot merge nobump
